rivertest/worker.go

```diff
@@ -0,0 +1,166 @@
+package rivertest
+
+import (
+	"context"
+	"encoding/json"
+	"sync/atomic"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+
+	"github.com/riverqueue/river"
+	"github.com/riverqueue/river/riverdriver"
+	"github.com/riverqueue/river/rivershared/util/ptrutil"
+	"github.com/riverqueue/river/rivershared/util/valutil"
+	"github.com/riverqueue/river/rivertype"
+)
+
+type Worker[T river.JobArgs, TTx any] struct {
+	client *river.Client[TTx]
+	config *river.Config
+	worker river.Worker[T]
+}
+
+func NewWorker[T river.JobArgs, TTx any](t *testing.T, driver riverdriver.Driver[TTx], config *river.Config, worker river.Worker[T]) *Worker[T, TTx] {
+	t.Helper()
+
+	config = config.WithDefaults()
+	client, err := river.NewClient(driver, config)
+	require.NoError(t, err)
+
+	return &Worker[T, TTx]{
+		client: client,
+		config: config,
+		worker: worker,
+	}
+}
+
+func (w *Worker[T, TTx]) Work(ctx context.Context, t *testing.T, args T) error {
+	t.Helper()
+	return w.WorkOpts(ctx, t, args, nil)
+}
+
+func (w *Worker[T, TTx]) WorkJob(ctx context.Context, t *testing.T, job *river.Job[T]) error {
+	t.Helper()
+	return w.WorkJobOpts(ctx, t, job, nil)
+}
+
+func (w *Worker[T, TTx]) WorkJobOpts(ctx context.Context, t *testing.T, job *river.Job[T], opts *river.InsertOpts) error {
+	t.Helper()
+
+	// populate river client into context:
+	ctx = WorkContext(ctx, w.client)
+
+	doInner := w.buildMiddlewareChain(job)
+	return doInner(ctx)
+}
+
+func (w *Worker[T, TTx]) WorkOpts(ctx context.Context, t *testing.T, args T, opts *river.InsertOpts) error {
+	t.Helper()
+
+	job, err := makeJobFromArgs(t, w.config, args, opts)
+	if err != nil {
+		return err
+	}
+
+	return w.WorkJobOpts(ctx, t, job, opts)
+}
+
+func (w *Worker[T, TTx]) buildMiddlewareChain(job *river.Job[T]) func(ctx context.Context) error {
+	// apply global middleware from client, as well as worker-specific middleware
+	workerMiddleware := w.worker.Middleware(job)
+
+	doInner := func(ctx context.Context) error {
+		jobTimeout := w.worker.Timeout(job)
+		if jobTimeout == 0 {
+			jobTimeout = w.config.JobTimeout
+		}
+
+		// No timeout if a -1 was specified.
+		// TODO: is this timeout being applied in the wrong place? Should it apply
+		// on the *outside* of all middleware?
+		if jobTimeout > 0 {
+			var cancel context.CancelFunc
+			ctx, cancel = context.WithTimeout(ctx, jobTimeout)
+			defer cancel()
+		}
+
+		return w.worker.Work(ctx, job)
+	}
```

Comment:
Would definitely be nice to be able to share all this with the main package a little better. I could see it being an easy mistake to make to add something in one place and then forget to do the other.